### PR TITLE
introducing global serializer

### PR DESCRIPTION
**Pull request by PavelPolyakov**
Resolves
https://github.com/pinojs/pino/issues/362

Please, let me know if I missed anything

### CODE REVIEW

File: docs/API.md [hunk 1]
```diff
@@ -45,6 +45,9 @@
     of objects. These functions should return an JSONifiable object and they
     should never throw. When logging an object, each top-level property matching the exact key of a serializer
     will be serialized using the defined serializer.
+    
+    **USE ON YOUR OWN RISK**
```
Comment (jsumners) [inline docs/API.md]: Why this warning?

File: docs/API.md [hunk 2]
```diff
@@ -45,6 +45,9 @@
...
+    As an exception it is possible to register `Symbol.for('*')` serializer, which accepts the complete log object (`pid`, `hostname` and all custom properties) and returns the updated complete log object.
```
Comment (jsumners) [inline docs/API.md]: "Alternatively, it is possible to register a serializer under the key `Symbol.for('*')` which will act upon the complete log object, i.e. every property."

Comment (PavelPolyakov) [reply docs/API.md]: updated

**Comment by davidmarkclements** [comment]
Symbol.for is a global namespace - over time collisions (especially against something like *) willl become likely. In general we should prefix Symbol.for namespaces with `pino.` - eg `Symbol.for('pino.*')`

Also could you please post before/after benchmark results here

**Comment by PavelPolyakov** [comment]
I used `Symbol.for('*')` alike the [metadata one's usage](https://github.com/pinojs/pino/blob/master/docs/API.md#metadata) . I think probability that anyone would have `Symbol.for('*')` in the object they are going to serialise is the same as having `Symbol.for('pino.*')`.

Here are the results on my MacPro mid 2015:
`master`: https://pastebin.com/tzjKiwv7
`feature/global-serializer`: https://pastebin.com/2bDWnrmM

**Comment by davidmarkclements** [comment]
Yeah see I disagree with metadata usage – I think it needs to be prefixed as well. 

@mcollina @jsumners – agree/disagree? 

Re benchmarks they look good!

**Comment by jsumners** [comment]
I'm good with namespacing the symbols.

**Comment by PavelPolyakov** [comment]
@davidmarkclements @jsumners 
I can bring the namespacing as part of this PR as well, but it'd mean that we need to bump the version (from 4 to 5), since it'll not be backward compatible.

**Comment by PavelPolyakov** [comment]
if you want me to do it, please, approve that this can be done in scope of this PR.

We also can somehow make the `Symbol.for('needsMetadata')` usage deprecated, till the next version is not released. BUT I don't have any good idea how to do that, since I assume we can not write any deprecation message to the console. Maybe just deprecate it through the documentation?

**Comment by mcollina** [comment]
I think we should leave `Symbol.for('needsMedata')` as is, and then namespace the new one. So we can release this as a minor. Also, open an issue about namespacing `Symbol.for('needsMedata')` so that we remember to do it for v5.

**Comment by PavelPolyakov** [comment]
Good, I've updated the new `Symbol`

**Comment by mcollina** [comment]
@jsumners are you good with this?

**Comment by mcollina** [comment]
Thanks for all the hard work @PavelPolyakov!

Landing now.

**Comment by grantcarthew** [comment]
For reference: this PR is a solution to this issue https://github.com/pinojs/pino/issues/355.

**Pull request merged**